### PR TITLE
*: add region statistics

### PR DESCRIPTION
**Pull request by nolouch**
ref: https://github.com/pingcap/pd/issues/856

### CODE REVIEW

File: server/handler.go [hunk 1]
```diff
@@ -397,3 +397,35 @@ func (h *Handler) AddRemovePeerOperator(regionID uint64, fromStoreID uint64) err
 	c.addOperator(op)
 	return nil
 }
+
+// GetDownPeerRegions gets the region with down peer.
+func (h *Handler) GetDownPeerRegions() map[string]map[uint64]*core.RegionInfo {
+	if h.s.cluster.regionStats != nil {
+		return h.s.cluster.regionStats.downPeers
```
Comment (disksing) [inline server/handler.go]: Should not read `cluster.regionStats` with out any mutex protected.

File: server/cluster.go [hunk 2]
```diff
@@ -452,6 +453,13 @@ func (c *RaftCluster) collectMetrics() {
 	}
 	statsMap.Collect()
 
+	regionStats := newRegionStatistics(c.cachedCluster.opt, c.coordinator.classifier)
+	for _, r := range cluster.GetRegions() {
+		regionStats.Observe(r)
+	}
```
Comment (disksing) [inline server/cluster.go]: Can you test or evaluate the performance cost here if we have 1 million regions?

File: pdctl/command/region_command.go [hunk 3]
```diff
@@ -183,3 +185,31 @@ func decodeProtobufText(text string) (string, error) {
 	}
 	return string(buf), nil
 }
+
+// NewRegionWithCheckCommand return a region with check subcommand of regionCmd
+func NewRegionWithCheckCommand() *cobra.Command {
+	r := &cobra.Command{
+		Use:   "check [miss|more|down|pending] <key>",
```
Comment (Connor1996) [inline pdctl/command/region_command.go]: What is "<key>" used for

Comment (nolouch) [reply pdctl/command/region_command.go]: useless.I forget to delete it

File: server/cluster.go (hunk 2 above)
Comment (nolouch) [reply server/cluster.go]: I need change to another method, GetRegions got poor performance.

File: server/coordinator.go [hunk 4]
```diff
@@ -114,6 +116,8 @@ func (c *coordinator) dispatch(region *core.RegionInfo) {
 	if op := c.replicaChecker.Check(region); op != nil {
 		c.addOperator(op)
 	}
+
+	c.regionStats.Observe(region)
```
Comment (disksing) [inline server/coordinator.go]: No Lock?

File: server/handler.go [hunk 5]
```diff
@@ -397,3 +400,71 @@ func (h *Handler) AddRemovePeerOperator(regionID uint64, fromStoreID uint64) err
...
+		return
+	}
+	cluster := c.cachedCluster
+	if cluster.regionStats == nil {
+		rd.JSON(w, http.StatusInternalServerError, errRegionStatsNotFound)
+		return
+	}
+	cluster.RLock()
+	rd.JSON(w, http.StatusOK, cluster.regionStats.stats[downPeer])
+	c.RUnlock()
```
Comment (Connor1996) [inline server/handler.go]: lock and unlock unpaired.

File: server/handler.go [hunk 6]
```diff
@@ -397,3 +400,71 @@ func (h *Handler) AddRemovePeerOperator(regionID uint64, fromStoreID uint64) err
...
+		rd.JSON(w, http.StatusInternalServerError, errNotBootstrapped)
+		return
+	}
+	cluster := c.cachedCluster
+	if cluster.regionStats == nil {
+		rd.JSON(w, http.StatusInternalServerError, errRegionStatsNotFound)
+		return
+	}
+	cluster.RLock()
+	rd.JSON(w, http.StatusOK, cluster.regionStats.stats[downPeer])
```
Comment (Connor1996) [inline server/handler.go]: I prefer add function getXXX() in clusterInfo and manage lock by clusterInfo itself.

File: server/region_statistics.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	r.stats[pendingPeer] = make(map[string]*core.RegionInfo)
+	r.stats[incorrectNameSpace] = make(map[string]*core.RegionInfo)
+	return r
+}
+
+func (r *regionStatistics) putRegion(typ regionStatisticType, namespace string, region *core.RegionInfo) {
+	regionID := region.GetId()
+	key := fmt.Sprintf("%s-%d", namespace, regionID)
+	r.stats[typ][key] = region
+
```
Comment (Connor1996) [inline server/region_statistics.go]: please remove blank line

File: server/region_statistics.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+		deleteIndex = oldIndex &^ peerTypeIndex
+	}
+	r.deleteEntry(deleteIndex, namespace, region.GetId())
+	r.index[region.GetId()] = peerTypeIndex
+}
+
+func (r *regionStatistics) Collect() {
+	metrics := make(map[string]map[string]float64)
+	for key := range r.stats[missPeer] {
+		addMetrics(metrics, key, "more_peer_region_count")
```
Comment (Connor1996) [inline server/region_statistics.go]: more -> miss

File: server/region_statistics.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+	// Region state.
+	namespace := r.classifier.GetRegionNamespace(region)
+	var (
+		peerTypeIndex regionStatisticType
+		deleteIndex   regionStatisticType
+	)
+	if len(region.Peers) < r.opt.GetMaxReplicas(namespace) {
+		r.putRegion(missPeer, namespace, region)
+		peerTypeIndex |= missPeer
+
```
Comment (Connor1996) [inline server/region_statistics.go]: please remove blank line

File: pdctl/command/region_command.go [hunk 10]
```diff
@@ -183,3 +185,31 @@ func decodeProtobufText(text string) (string, error) {
...
+	state := args[0]
+
+	prefix := regionsCheckPrefix + "/" + state
+	r, err := doRequest(cmd, prefix, http.MethodGet)
+	if err != nil {
+		fmt.Printf("Failed to get region: %s\n", err)
+		return
+	}
+	fmt.Println(r)
+
```
Comment (Connor1996) [inline pdctl/command/region_command.go]: Please remove blank line

File: server/region_statistics.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	if len(region.DownPeers) > 0 {
+		r.putRegion(downPeer, namespace, region)
+		peerTypeIndex |= downPeer
+	}
+	if len(region.PendingPeers) > 0 {
+		r.putRegion(pendingPeer, namespace, region)
+		peerTypeIndex |= pendingPeer
+	}
+	oldIndex, ok := r.index[region.GetId()]
+	if ok {
```
Comment (Connor1996) [inline server/region_statistics.go]: -> if oldIndex, ok := r.index[region.GetId()]; ok {

File: server/handler.go [hunk 12]
```diff
@@ -397,3 +400,71 @@ func (h *Handler) AddRemovePeerOperator(regionID uint64, fromStoreID uint64) err
 	c.addOperator(op)
 	return nil
 }
+
+// ResponseDownPeerRegions gets the region with down peer.
+func (h *Handler) ResponseDownPeerRegions(w http.ResponseWriter, rd *render.Render) {
```
Comment (Connor1996) [inline server/handler.go]: I prefer extract the common part

File: server/api/region.go [hunk 13]
```diff
@@ -142,6 +142,51 @@ func (h *regionsHandler) GetAll(w http.ResponseWriter, r *http.Request) {
...
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+	}
+	h.rd.JSON(w, http.StatusOK, res)
+}
+
+func (h *regionsHandler) GetMorePeerRegions(w http.ResponseWriter, r *http.Request) {
+	handler := h.svr.GetHandler()
+	res, err := handler.GetMorePeerRegions()
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
```
Comment (Connor1996) [inline server/api/region.go]: you lost return

File: server/api/region.go [hunk 14]
```diff
@@ -142,6 +142,51 @@ func (h *regionsHandler) GetAll(w http.ResponseWriter, r *http.Request) {
...
+	}
+	h.rd.JSON(w, http.StatusOK, res)
+}
+
+func (h *regionsHandler) GetPendingPeerRegions(w http.ResponseWriter, r *http.Request) {
+	handler := h.svr.GetHandler()
+	res, err := handler.GetPendingPeerRegions()
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
```
Comment (Connor1996) [inline server/api/region.go]: ditto

File: server/api/region.go [hunk 15]
```diff
@@ -142,6 +142,51 @@ func (h *regionsHandler) GetAll(w http.ResponseWriter, r *http.Request) {
...
+	}
+	h.rd.JSON(w, http.StatusOK, res)
+}
+
+func (h *regionsHandler) GetDownPeerRegions(w http.ResponseWriter, r *http.Request) {
+	handler := h.svr.GetHandler()
+	res, err := handler.GetDownPeerRegions()
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
```
Comment (Connor1996) [inline server/api/region.go]: ditto

File: server/api/region.go [hunk 16]
```diff
@@ -142,6 +142,51 @@ func (h *regionsHandler) GetAll(w http.ResponseWriter, r *http.Request) {
...
+	}
+	h.rd.JSON(w, http.StatusOK, res)
+}
+
+func (h *regionsHandler) GetIncorrectNamespaceRegions(w http.ResponseWriter, r *http.Request) {
+	handler := h.svr.GetHandler()
+	res, err := handler.GetIncorrectNamespaceRegions()
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
```
Comment (Connor1996) [inline server/api/region.go]: ditto

File: server/region_statistics_test.go [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+	downPeers := []*pdpb.PeerStats{
+		{Peer: peers[0], DownSeconds: 3608},
+		{Peer: peers[1], DownSeconds: 3608},
+	}
+	r1 := &metapb.Region{Id: 1, Peers: peers, StartKey: []byte("aa"), EndKey: []byte("bb")}
+	r2 := &metapb.Region{Id: 2, Peers: peers[0:2], StartKey: []byte("cc"), EndKey: []byte("dd")}
+	region1 := core.NewRegionInfo(r1, peers[0])
+	region2 := core.NewRegionInfo(r2, peers[0])
+	regionStats := newRegionStatistics(opt, namespace.DefaultClassifier)
```
Comment (Connor1996) [inline server/region_statistics_test.go]: not cover incorrect namespace case

File: server/cache.go [hunk 18]
```diff
@@ -461,6 +470,11 @@ func (c *clusterInfo) handleRegionHeartbeat(region *core.RegionInfo) error {
 			c.updateStoreStatus(p.GetStoreId())
 		}
 	}
+
+	if c.regionStats != nil {
```
Comment (disksing) [inline server/cache.go]: I think _c.regionStats_ won't be nil.

File: server/cluster.go [hunk 19]
```diff
@@ -109,6 +109,8 @@ func (c *RaftCluster) start() error {
 	}
 	c.cachedCluster = cluster
 	c.coordinator = newCoordinator(c.cachedCluster, c.s.hbStreams, c.s.classifier)
+	regionStats := newRegionStatistics(c.s.scheduleOpt, c.s.classifier)
+	c.cachedCluster.bindRegionStatistics(regionStats)
```
Comment (disksing) [inline server/cluster.go]: I think you can just `c.cachedCluster.regionStats = newRegionStatistics(c.s.scheduleOpt, c.s.classifier)`.

File: server/cache.go [hunk 20]
```diff
@@ -479,6 +493,25 @@ func (c *clusterInfo) handleRegionHeartbeat(region *core.RegionInfo) error {
...
+func (c *clusterInfo) GetRegionStatsByType(typ regionStatisticType) map[string]*core.RegionInfo {
+	if c.regionStats == nil {
+		return nil
+	}
+	res := make(map[string]*core.RegionInfo)
+	c.RLock()
+	defer c.RUnlock()
+	for k, r := range c.regionStats.stats[typ] {
+		res[k] = r.Clone()
+	}
```
Comment (disksing) [inline server/cache.go]: Why not make this a method of `RegionStats`?

File: server/region_statistics.go [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+		stats[namespace] = mm
+		mm[typ] = 1
+	} else {
+		m[typ]++
+	}
+}
+
+func (r *regionStatistics) deleteEntry(deleteIndex regionStatisticType, namespace string, regionID uint64) {
+	index := regionStatisticType(1)
+	key := fmt.Sprintf("%s-%d", namespace, regionID)
```
Comment (disksing) [inline server/region_statistics.go]: The namespace of a region could change.

File: server/cache.go (hunk 18 above)
Comment (nolouch) [reply server/cache.go]: In test will be nil.

File: pdctl/command/region_command.go [hunk 22]
```diff
@@ -183,3 +185,28 @@ func decodeProtobufText(text string) (string, error) {
 	}
 	return string(buf), nil
 }
+
+// NewRegionWithCheckCommand return a region with check subcommand of regionCmd
+func NewRegionWithCheckCommand() *cobra.Command {
+	r := &cobra.Command{
+		Use:   "check [miss-replica|more-replica|down-replica|pending-replica|incorrect-ns]",
```
Comment (disksing) [inline pdctl/command/region_command.go]: more -> extra.

File: server/region_statistics.go [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+func (r *regionStatistics) deleteEntry(deleteIndex regionStatisticType, regionID uint64) {
+	index := regionStatisticType(1)
+	for deleteIndex > 0 {
+		needDelete := deleteIndex & 1
+		if needDelete == 1 {
+			delete(r.stats[index], regionID)
+		}
+		deleteIndex = deleteIndex >> 1
+		index = index << 1
+	}
```
Comment (disksing) [inline server/region_statistics.go]: :smile: Try this:
```
for typ := 1; typ <= maxStatisticsType; typ <<= 1 {
  if deleteIndex & typ != 0 {
    // ...
  }
}
```

File: server/region_statistics.go [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+func (r *regionStatistics) Collect() {
+	metrics := make(map[string]float64)
+	metrics["miss_peer_region_count"] = float64(len(r.stats[missPeer]))
+	metrics["extra_peer_region_count"] = float64(len(r.stats[extraPeer]))
+	metrics["down_peer_region_count"] = float64(len(r.stats[downPeer]))
+	metrics["pending_peer_region_count"] = float64(len(r.stats[pendingPeer]))
+	metrics["incorrect_namespace_region_count"] = float64(len(r.stats[incorrectNamespace]))
+	for label, value := range metrics {
+		regionStatusGauge.WithLabelValues(label).Set(value)
+	}
```
Comment (disksing) [inline server/region_statistics.go]: The map is not needed. Just set each label is more simple.

**Comment by disksing** [comment]
LGTM. PTAL @Connor1996

**Pull request merged**